hazelcast/src/main/java/com/hazelcast/client/impl/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -185,134 +224,299 @@ protected NioNetworking initNetworking(final HazelcastClientInstanceImpl client)
+                // Otherwise we can get startup delays.
+                Collections.shuffle(result.primary());
+                Collections.shuffle(result.secondary());
+            }
+            providedAddresses.addAll(result.primary());
+            providedAddresses.addAll(result.secondary());
+        } catch (NullPointerException e) {
+            throw e;
+        } catch (Exception e) {
+            logger.warning("Exception from AddressProvider: " + discoveryService, e);
+        }
+
+        addresses.addAll(providedAddresses);
+
+        return addresses;
     }
 
-    private void checkAllowed(Address target) throws IOException {
-        if (!alive) {
-            throw new HazelcastClientNotActiveException("ConnectionManager is not active!");
+    private static <T> Set<T> shuffle(Set<T> set) {
+        List<T> shuffleMe = new ArrayList<T>(set);
+        Collections.shuffle(shuffleMe);
+        return new LinkedHashSet<T>(shuffleMe);
+    }
+
+    private void triggerReconnectToCluster() {
+        if (reconnectMode == OFF) {
+            logger.info("RECONNECT MODE is off. Shutting down the client");
+            shutdownWithExternalThread();
+            return;
         }
-        connectionStrategy.beforeGetConnection(target);
+        if (client.getLifecycleService().isRunning()) {
+            try {
+                connectToClusterAsync();
+            } catch (RejectedExecutionException r) {
+                shutdownWithExternalThread();
+            }
+        }
+    }
+
+    private void shutdownWithExternalThread() {
+        new Thread(() -> {
+            try {
+                client.getLifecycleService().shutdown();
+            } catch (Exception exception) {
+                logger.severe("Exception during client shutdown ", exception);
+            }
+        }, client.getName() + ".clientShutdown-").start();
+    }
+
+    @Override
+    public Collection<ClientConnection> getActiveConnections() {
+        return activeConnections.values();
+    }
+
+    @Override
+    public boolean isAlive() {
+        return isAlive.get();
+    }
+
+    @Override
+    public UUID getClientUuid() {
+        return clientUuid;
     }
 
+    @Override
+    public Connection getConnection(@Nonnull Address target) {
+        checkClientActive();
+        return activeConnections.get(inetSocketAddressCache.get(target));
+    }
 
-    private AuthenticationFuture triggerConnect(Address target) {
-        connectionStrategy.beforeOpenConnection(target);
+    @Override
+    public Connection getOrConnect(@Nonnull Address address) {
+        checkClientActive();
+        InetSocketAddress inetSocketAddress = inetSocketAddressCache.get(address);
+        ClientConnection connection = activeConnections.get(inetSocketAddress);
+        if (connection != null) {
+            return connection;
+        }
+        synchronized (locks[abs(inetSocketAddress.hashCode() % CONCURRENT_AUTH_COUNT)]) {
```

Comment:
I want to prevent opening more than one connection to the same address concurrently. Can I achieve this using the`atomic counter`? How?

Reply:
Ok, I understand the need now. You need a lock but I still think that we do not need to limit the number of locks. I believe it can be made unlimited. A wild suggestion: Is the address object uniquely created for a certain address, and hence single instance? If so, can directly call `synchronized` on address? (We do not even need the lock array)